Widgets: use Photon to display image in Image Widget.

eliorivero opened the pull request:
Fixes #4141
#### Changes proposed in this Pull Request:
- image is loaded through Photon
#### Changelog

Extra Widgets: if Photon is available, use it to serve the image in the Image widget.

jeherve commented on modules/widgets/image-widget.php:
```diff
@@ -70,7 +70,10 @@ public function widget( $args, $instance ) {
 		if ( '' != $instance['img_url'] ) {
 
-			$output = '<img src="' . esc_attr( $instance['img_url'] ) .'" ';
+			$output = '<img src="' . jetpack_photon_url( esc_attr( $instance['img_url'] ), array(
```
Shouldn't it be `esc_url` instead of `esc_attr`?

bhubbard replied:
@eliorivero Can you please update this pull request to use `esc_url()`

jeherve commented:
I just tested it with an old image widget that was already active on my site, and for some reason `width` and `height` were set to 0.:

`<img src="https://i2.wp.com/i.cloudup.com/uUKVEHkIBe.gif?w=0&amp;h=0&amp;quality=80&amp;strip=all&amp;ssl=1" class="alignnone">`

I can see `width` and `height` values in the widget settings, though.

![screen shot 2016-10-27 at 21 27 58](https://cloud.githubusercontent.com/assets/426388/19782027/4ad32ee8-9c8c-11e6-815d-8ccba69e127c.png)

No problem when taking the same image and creating a new Image widget, though:

`<img src="https://i2.wp.com/i.cloudup.com/uUKVEHkIBe.gif?w=400&amp;h=225&amp;quality=80&amp;strip=all&amp;ssl=1" class="alignnone" width="400" height="225">`

I wonder if we should add a check like [here](https://github.com/Automattic/jetpack/blob/e868342b5967441ec933f19f7272e900f02a309a/modules/widgets/image-widget.php#L87) to cover old widgets like that?

lancewillett commented:
Needs @eliorivero author reply before proceeding.

kraftbj commented:
Do we want to always use Photon for this widget or only when the Photon module is enabled?

lancewillett commented:
@kraftbj Based on the [original request](https://github.com/Automattic/jetpack/issues/4141), only when the module is enabled.

kraftbj commented on modules/widgets/image-widget.php:
```diff
@@ -70,7 +70,10 @@ public function widget( $args, $instance ) {
 		if ( '' != $instance['img_url'] ) {
 
-			$output = '<img src="' . esc_attr( $instance['img_url'] ) .'" ';
+			$output = '<img src="' . esc_url( jetpack_photon_url( $instance['img_url'], array(
```
Need to add `( method_exists( 'Jetpack', 'is_module_active' ) && Jetpack::is_module_active( 'photon' ) )` conditional.

eliorivero replied:
I added only `Jetpack::is_module_active( 'photon' )`. Note that in many widgets (like Follow Button) we call `Jetpack::is_active()` on `widgets_init` and only then the widget is registered so by the time of widget registration the `Jetpack` class (and its methods) exists.

eliorivero commented:
@jeherve I can't reproduce the issue. Starting with master and Photon enabled, I added an Image widget an image URL and 400 x 225 like you have. Saved, checked in front end, all good. Switched to branch, checked in front end again, width and height were there.
That said, when you point this line `if ( '' != $instance['img_width'] ) {` I'm not really sure which condition you'd like to add to "cover old widgets". If you have something on your mind feel free to go ahead and add it.

jeherve commented:
I'm afraid I can still experience an issue, but it now happens to new widgets, not existing ones. Try the following:

1. Check out your branch.
2. Go to Appearance > Widgets.
3. Drag a new Image Widget into a widget area.
4. Paste this URL in the URL field: `http://i.wpne.ws/igUH/Screen%20Shot%202017-01-09%20at%2011.26.00.png`
5. Hit the Save button at the bottom of the widget settings.
6. Load your site's front page.
7. The image is displayed, but no width or height parameters are added to the URL or to the image tag.

This doesn't happen with image widgets that were created before the branch was checked.

Compare this image widget, added before to check out the branch:

![screen shot 2017-01-09 at 14 42 36](https://cloud.githubusercontent.com/assets/426388/21768416/e6b7bd02-d679-11e6-8e68-713464f6000c.png)

And this image widget, added after checking out the branch:

![screen shot 2017-01-09 at 14 43 16](https://cloud.githubusercontent.com/assets/426388/21768434/f9fd0ade-d679-11e6-9e3d-fa551e4103db.png)

Can you reproduce as well?

eliorivero commented:
I followed your steps and found that it only happens in my installation when Photon module is deactivated:

<img width="529" alt="captura de pantalla 2017-02-06 a las 19 18 35" src="https://cloud.githubusercontent.com/assets/1041600/22668917/3b95d52e-eca1-11e6-9df1-83f62600ac81.png">

If it's active, the dimensions are properly set
<img width="528" alt="captura de pantalla 2017-02-06 a las 19 19 30" src="https://cloud.githubusercontent.com/assets/1041600/22668918/3b9da97a-eca1-11e6-88b7-622aac61a78b.png">
